### ISSUE TITLE
Treating non-functional positioners as sky

### ISSUE THREAD

**Issue by tskisner** [desihub/fiberassign#324]
Positioners that are not moveable are often pointing at empty sky, and it would be useful to treat them as such for our sky budget.  

One possibility is to internally count each stuck positioner as 0.5 sky fiber in the per-petal counts.  Another possibility would be to pre-check the stuck RA/DEC location of stuck positioners on every tile and create a lookup of whether each of these is sky or not.  This check would call an external python function, but the data would need to be stored in C++ for use during assignment.

**Comment by tskisner** [comment]
Note that the resolution of this is coupled to #214, since without that, a disabled canbus could result in a local excess of sky fibers.

**Comment by dstndstn** [comment]
Hi Ted, I'm interested in trying to tackle this.  I have only just started looking at the fiberassign code, but I'm wondering if an approach could be:
- for each tile, find the stuck positioners and do the lookup for whether they will be in good sky locations for that tile center.  Add SKY targets for those locations -- so that they just *happen* to be reachable by the stuck positioners
- assign the stuck positioners first (before the first round of science fibers), so that those that can reach SKY targets do so

Or is that trying to be too sneaky, and instead I should just do the SKY lookup for stuck positioners when assigning them?

I'm guessing that #214 is going to involve changes to the forced-assignment of sky targets step, so in implementation might be basically orthogonal.

**Comment by tskisner** [comment]
It is true that this work is mostly orthogonal with #214, it is just that we would not deploy the work discussed in this issue until that change is made.

There are several aspects to implementing this:

- We do not need to manipulate the input target information.  Instead, we are just tracking for each tile and each stuck positioner, is the corresponding RA/DEC a sky or not?  This involves calling your external python code and building up that information.
- The result of this is passed to the C++ `Assignment` class and in the constructor it will be used to initialize the internal counters for sky per petal.
- Everything past that point should work as it is.  Later, when processing the data, the locations of stuck positioners can be interpreted as sky or not by calling the same external python code.

So basically, we are not creating new fake sky targets.  We are just deciding whether to treat stuck positioners as sky for each tile.  Do you already have the external python tools that can return a sky / no-sky value for a given tile center and location in RA / DEC?

If you have time to work on this I can scope it out in more detail, but perhaps we should coordinate on the call tomorrow and come up with a plan.

**Comment by dstndstn** [comment]
Hi Ted,
Yes, the code for looking up an RA,Dec in the sky images exists (and is pretty trivial).
This implementation makes sense.  I'll see what I can figure out.  I'm just getting my bearings in the code.
thanks!

**Comment by dstndstn** [comment]
Poking in the code a bit, you're suggesting pre-loading Assignment's `nassign_petal[TARGET_TYPE_SKY]` (which I guess will become `nassign_casette` or something...) with the number of stuck positioners that happen to land on a good sky location for each tile?

**Comment by tskisner** [comment]
That is the basic idea.  The challenge is that the logical place for that to happen would be in the `Assignment` constructor which is C++, but we need to call an external python function.  So instead, I would modify the constructor:

https://github.com/desihub/fiberassign/blob/dced43beebe4c513bde055900ff06a696f4c73dc/src/assign.cpp#L17

To take a `std::map < std::map <int32_t, bool> > stuck_sky` argument.  Then in this loop:

https://github.com/desihub/fiberassign/blob/dced43beebe4c513bde055900ff06a696f4c73dc/src/assign.cpp#L51

If `tgtype` is `TARGET_TYPE_SKY`, then for each tile you can iterate over the locations in `stuck_sky`, get the petal number for that location with `hw_->loc_petal[loc]`, and if that location is sky, increment the counter for that petal.

Then modify the bindings here:

https://github.com/desihub/fiberassign/blob/dced43beebe4c513bde055900ff06a696f4c73dc/src/_pyfiberassign.cpp#L1506

To accept a `std::map<std::map<int32_t, bool>>`.  Next, somewhere just before here:

https://github.com/desihub/fiberassign/blob/dced43beebe4c513bde055900ff06a696f4c73dc/py/fiberassign/scripts/assign.py#L360

Create a new python function that takes a `Tiles` object and a `Hardware` object and returns a nested dictionary keyed on tile, then location and with values of bool.  This python function can then get the RA / DEC of all stuck positioners for each tile and find if they are sky:

```
# Written on the fly- not tested!   Assume a tiles and hw input.

stuck_sky = dict()
for tile_id, tile_ra, tile_dec, tile_theta in zip(tiles.id, tiles.ra, tiles.dec, tiles.theta):
    stuck_sky[tile_id] = dict()

    # Stuck locations and their angles
    stuck_loc = [x for x in hw.locations if hw.state[x] & FIBER_STATE_STUCK]
    stuck_theta = [hw.loc_theta_pos[x] for x in stuck_loc]
    stuck_phi = [hw.loc_phi_pos[x] for x in stuck_loc]

    # Convert positioner angle orientations to curved focal surface X / Y (not CS5)
    # Note:  we could add some methods to the python bindings to vectorize this or make it less clunky...
    stuck_x = list()
    stuck_y = list()
    for loc, theta, phi in zip(stuck_loc, stuck_theta, stuck_phi):
        loc_x, loc_y = hw.thetaphi_to_xy(
            hw.loc_pos_curved_mm[loc],
            theta,
            phi,
            hw.loc_theta_arm[loc],
            hw.loc_phi_arm[loc],
            hw.loc_theta_offset[loc],
            hw.loc_phi_offset[loc],
            hw.loc_theta_min[loc],
            hw.loc_phi_min[loc],
            hw.loc_theta_max[loc],
            hw.loc_phi_max[loc]
        )
        stuck_x.append(loc_x)
        stuck_y.append(loc_y)
    stuck_x = np.array(stuck_x)
    stuck_y = np.array(stuck_y)
    
    loc_ra, loc_dec = hw.xy2radec_multi(
        tile_ra,
        tile_dec,
        tile_theta,
        stuck_x,
        stuck_y,
        False,
        0
    )

    for loc, ra, dec in zip(stuck_loc, loc_ra, loc_dec):
        # Call an external function- is this RA / DEC a sky?
        if yes:
            stuck_sky[tile_id][loc] = True
        else:
            stuck_sky[tile_id][loc] = False
```

Then you can pass the resulting nested dictionary to the `Assignment` constructor (pybind11 copies it under the hood to the nested `std::map`).

**Comment by dstndstn** [comment]
Fantastic, I love it when people do my homework for me ;)

**Comment by dstndstn** [comment]
PS, would you ever consider typedef'ing tileid, locid, targetid types?  Currently there are lots of `map<int32_t, ...>`s floating around where it's not entirely obvious at first glance what the indices are.  I guess they're actually all of the form `[tileid][locid]`, so it's not as ambiguous as it might seem on the surface.  And the concreteness of int32 simplifies understanding the code.

**Comment by dstndstn** [comment]
I think I've got this working!  On SV3 tile 190 test,
415 of 567 stuck positioners land on good sky

Before: 477 sky 3794 science
After: 157 sky 4116 science

One question: it seems like it would be useful to track in the output table the stuck fibers we thought would land on good sky.  Is there a bit I can latch on to for that?

**Comment by djschlegel** [comment]
AWESOME!   The question of how to make these identifiable in the fiberassign files is a good one.  It may be enough to assign these a negative TARGETID (as we do for all non-working positioners) but set OBJTYPE='SKY'.

**Comment by tskisner** [comment]
Yes, please let's use the existing columns if possible and not change the format again...  I like @djschlegel 's suggestion.  Also there was some commentary on #214 that was more relevant to this issue.  Essentially about whether we should treat these "accidental skies" as "real" skies or something else (e.g. "half a sky").

Originally I was against this, but there are two factors that would make this kind of thing sensible:
1.  In the `Assign` constructor, when taking the "stuck_sky" dictionary and populating the assignment counters, it would be easy to just take the total per petal / cassette / slitblock and divide by 2.
2. For stuck positioners, I have noticed that the POS_T / POS_P angles do "drift" from one night to the next.  Or at least the online system estimate of the positioner locations drifts.  So counting these as half a "real" sky would hedge against a positioner drifting towards a source in between when the tile was assigned and when it was observed.

**Comment by dstndstn** [comment]
Done in #337
